tests/functional/models/test_application.py

```diff
@@ -0,0 +1,145 @@
+
+        # found an application, it should return an application with matched name.
+        self.resin.models.application.create('FooBar', 'Raspberry Pi 2')
+        self.assertEqual(self.resin.models.application.get('FooBar')['app_name'], 'FooBar')
+
+    def test_has(self):
+        # should be true if the application name exists, otherwise it should return false.
+        self.resin.models.application.create('FooBar', 'Raspberry Pi 2')
+        self.assertFalse(self.resin.models.application.has('FooBar1'))
+        self.assertTrue(self.resin.models.application.has('FooBar'))
+
+    def test_has_any(self):
+        # given no applications, it should return false.
+        self.assertFalse(self.resin.models.application.has_any())
+
+        # should return true if at least one application exists.
+        self.resin.models.application.create('FooBar', 'Raspberry Pi 2')
+        self.assertTrue(self.resin.models.application.has_any())
+
+    def test_get_by_id(self):
+        # raise resin.exceptions.ApplicationNotFound if no application found.
+        with self.assertRaises(self.helper.resin_exceptions.ApplicationNotFound):
+            self.resin.models.application.get_by_id(1)
+
+        # found an application, it should return an application with matched id.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        self.assertEqual(self.resin.models.application.get_by_id(app['id'])['id'], app['id'])
+
+    def test_remove(self):
+        # should be able to remove an existing application by name.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        self.assertEqual(len(self.resin.models.application.get_all()), 1)
+        self.resin.models.application.remove('FooBar')
+        self.assertEqual(len(self.resin.models.application.get_all()), 0)
+
+    def test_generate_provisioning_key(self):
+        # should be rejected if the application id does not exist.
+        with self.assertRaises(self.helper.resin_exceptions.ApplicationNotFound):
+            self.resin.models.application.generate_provisioning_key('5685')
+
+        # should be able to generate a provisioning key by app id.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        key = self.resin.models.application.generate_provisioning_key(app['id'])
+        self.assertEqual(len(key), 32)
+
+    def test_enable_rolling_updates(self):
+        # should enable rolling update for the applications devices.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        self.resin.models.application.enable_rolling_updates(app['id'])
+        app = self.resin.models.application.get('FooBar')
+        self.assertTrue(app['should_track_latest_release'])
+
+    def test_disable_rolling_updates(self):
```

Comment:
With all these enable and disable tests, I wonder if it _might_ make more sense to do e.g.
```python
    def test_rolling_updates(self): # or test_toggle_rolling_updates
        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
        # I guess at this point you could also test the default value of app['should_track_latest_release'] ?
        # should enable rolling update for the applications devices.
        self.resin.models.application.enable_rolling_updates(app['id'])
        app = self.resin.models.application.get('FooBar')
        self.assertTrue(app['should_track_latest_release'])
        # should disable rolling update for the applications devices.
        self.resin.models.application.disable_rolling_updates(app['id'])
        app = self.resin.models.application.get('FooBar')
        self.assertFalse(app['should_track_latest_release'])
```
?
And if the default value of `app['should_track_latest_release']` is True, then it'd probably make sense to test disabling it before testing enabling it.

Reply:
Personally I'd rather have more & smaller tests - makes them easier to read, easier to tell what exactly is broken if they fail, and easier to run separate bits in isolation.

Reply:
Fair enough. Should there also be tests for the default values? And presumably if the default value is e.g. True, then when testing the "enable" function you'd first need to disable it, to ensure that the enable isn't just a NOP?

Reply:
@lurch, I don't think we need a test for default values here but it makes sense to enable/disable first before testing disable/enable function to ensure our test is doing a right thing.

Updated both test cases.